src/test/java/org/mule/service/http/impl/service/client/HttpClientStreamingTestCase.java

```diff
@@ -0,0 +1,272 @@
+  public void nonBlockingStreaming() throws Exception {
+    HttpClient client = service.getClientFactory().create(clientBuilder.build());
+    client.start();
+    final Reference<HttpResponse> responseReference = new Reference<>();
+    try {
+      client.send(getRequest(), RESPONSE_TIMEOUT, true, null, new ResponseHandler() {
+
+        @Override
+        public void onCompletion(HttpResponse response) {
+          responseReference.set(response);
+        }
+
+        @Override
+        public void onFailure(Exception exception) {
+          latch.release();
+          fail("Response should have been received despite latch being set.");
+        }
+      });
+      pollingProber.check(new ResponseReceivedProbe(responseReference));
+      verifyStreamed(responseReference.get());
+    } finally {
+      client.stop();
+    }
+  }
+
+  @Test
+  @Description("Uses a non streaming HTTP client to send a non blocking request which will not finish until the stream is released.")
+  public void nonBlockingMemory() throws Exception {
+    HttpClient client = service.getClientFactory().create(clientBuilder.setStreaming(false).build());
+    client.start();
+    final Reference<HttpResponse> responseReference = new Reference<>();
+    try {
+      client.send(getRequest(), RESPONSE_TIMEOUT, true, null, new ResponseHandler() {
+
+        @Override
+        public void onCompletion(HttpResponse response) {
+          responseReference.set(response);
+        }
+
+        @Override
+        public void onFailure(Exception exception) {
+          latch.release();
+          fail("Response should have been received despite latch being set.");
+        }
+      });
+      verifyNotStreamed(responseReference);
+    } finally {
+      client.stop();
+    }
+  }
+
+  @Test
+  @Description("Uses a streaming HTTP client to send a blocking request which will finish before the stream is released.")
+  public void blockingStreaming() throws Exception {
+    HttpClient client = service.getClientFactory().create(clientBuilder.build());
+    client.start();
+    try {
+      HttpResponse response = client.send(getRequest(), RESPONSE_TIMEOUT, true, null);
+      verifyStreamed(response);
+    } finally {
+      client.stop();
+    }
+  }
+
+  @Test
+  @Description("Uses a non streaming HTTP client to send a request which will not finish until the stream is released.")
+  public void blockingMemory() throws Exception {
+    HttpClient client = service.getClientFactory().create(clientBuilder.setStreaming(false).build());
+    client.start();
+    Reference<HttpResponse> responseReference = new Reference<>();
+    ExecutorService executorService = Executors.newSingleThreadExecutor();
+    try {
+      executorService.execute(() -> {
+        try {
+          responseReference.set(client.send(getRequest(), RESPONSE_TIMEOUT, true, null));
+        } catch (Exception e) {
+          fail("Client failed executing request.");
```

Comment:
Will this fail actually fail test, because it'll be executed by a different thread.